test/src/com/legrig/CPSNonCPSTestClassSpec.groovy

```diff
@@ -0,0 +1,85 @@
+package com.legrig
+
+import com.cloudbees.groovy.cps.impl.CpsCallableInvocation
+import support.jenkins.CPSSpecification
+
+class CPSNonCPSTestClassSpec extends CPSSpecification{
+
+  Class testSubjectClass = CPSNonCPSTestClass
+
+  def "CPSNonCPSTestClass.nonCpsMethod() Test"(){
+
+    when: 'we call the method'
+      def itWorks = sut.nonCpsMethod()
+
+    then: 'it works'
+      itWorks
+  }
+
+  def "CPSNonCPSTestClass.cpsMethod() from NonCPS Test"(){
+
+    when: 'we call method directly'
+      def itWorks = sut.cpsMethod()
+
+    then: 'we get an exception'
+      thrown CpsCallableInvocation
+
+    and: 'it does not work'
+      ! itWorks
+  }
+
+  def "CPSNonCPSTestClass.cpsMethod() from CPS Test"(){
+
+    when: 'we call method via invokeCPSMethod'
+      def itWorks = invokeCPSMethod(sut, 'cpsMethod')
+
+    then: 'it works'
+      itWorks
+  }
+
+  def "CPSNonCPSTestClass.cpsMethodCallingNonCps() from NonCPS Test"(){
+
+    when: 'we call method directly'
+      def itWorks = sut.cpsMethodCallingNonCps()
+
+    then: 'we get an exception'
+      thrown CpsCallableInvocation
+
+    and: 'it does not work'
+      ! itWorks
+  }
+
+  def "CPSNonCPSTestClass.cpsMethodCallingNonCps() from CPS Test"(){
+
+    when: 'we call method via invokeCPSMethod'
+      def itWorks = invokeCPSMethod(sut, 'cpsMethodCallingNonCps')
+
+    then: 'it works'
+      itWorks
+  }
+
+
+  def "CPSNonCPSTestClass.cpsMethodCallingNonCpsCallingCps() from NonCPS Test"(){
+
+    when: 'we call method directly'
+      def itWorks = sut.cpsMethodCallingNonCpsCallingCps()
+
+    then: 'we get an exception'
+      thrown CpsCallableInvocation
+
+    and: 'it does not work'
+      ! itWorks
+  }
+
+  def "CPSNonCPSTestClass.cpsMethodCallingNonCpsCallingCps() from CPS Test"(){
```

Comment:
I see when following test cases are missing. It fails for these two cases. 



when:  invokeCPSMethod(sut, "nonCpsMethod")
then:         thrown IllegalStateException;



when: invokeCPSMethod(sut, "nonCpsMethodCallingCpsMethod")
then:         thrown IllegalStateException;

Reply:
OK, I think we do not want those cases as it is not something you can do with code when you are using, so supporting may only confuse things.  You are intended to only run `invokeCPSmethod()` on CPS methods. It is explicitly set to throw an exception (`InvalidCPSInvocation`) if you call it on `NonCPS` code. 

I do not see it as a problem, since you know what you are testing and you would know if it is CPS or not and test acordingly. I can see why this would never be an issue in case of `asCPSScript()` since that is guaranteed to be CPS transformed.

Reply:
Also, if you really want to run whatever without knowing, there is also `safeInvokeMethod()` that will handle CPS or NonCPS without knowing - not too useful in context of testing, but it is there

Reply:
Hmm, does not look like either `safeInvokeMethod()` nor `isCPSTransformed()` work. Not sure why, will have to investigate.

Reply:
To be able to catch issues where "non cps code calls cps code" we would need support to test everything without knowing what kind it is. Ideally we want something like a compiler that detects any cps mismatch issues before we run the code in jenkins.

Reply:
All code in pipeline is CPS by default, so anything unknown is CPS, and anything you write, you know for sure.

Reply:
cps mismatch(https://www.jenkins.io/doc/book/pipeline/cps-method-mismatches/) happens because not all code is cps transformed. Eg) constructors are not cps transformed. Hence calling a setter (which is cps transformed) from a constructor will cause cps method mismatch. To ensure that our code works in jenkins we need a way to catch this.

Reply:
Correct. That is the core reason for this approach -  this will absolutely catch those cases as is.

Reply:
Added a few unit tests to demo it in #5